internal/envoy/v3/route.go

```diff
@@ -34,6 +34,96 @@ import (
 	"github.com/projectcontour/contour/internal/sorter"
 )
 
+// VirtualHostAndRoutes converts a DAG virtual host and routes to an Envoy virtual host.
+func VirtualHostAndRoutes(vh *dag.VirtualHost, dagRoutes []*dag.Route, secure bool, authService *dag.ExtensionCluster) *envoy_route_v3.VirtualHost {
+	var envoyRoutes []*envoy_route_v3.Route
+	for _, route := range dagRoutes {
+		envoyRoutes = append(envoyRoutes, Route(route, vh.Name, secure, authService))
+	}
+
+	evh := VirtualHost(vh.Name, envoyRoutes...)
+
+	if vh.CORSPolicy != nil {
+		evh.Cors = CORSPolicy(vh.CORSPolicy)
+	}
+	if vh.RateLimitPolicy != nil && vh.RateLimitPolicy.Local != nil {
+		if evh.TypedPerFilterConfig == nil {
+			evh.TypedPerFilterConfig = map[string]*any.Any{}
+		}
+		evh.TypedPerFilterConfig["envoy.filters.http.local_ratelimit"] = LocalRateLimitConfig(vh.RateLimitPolicy.Local, "vhost."+vh.Name)
+	}
+
+	if vh.RateLimitPolicy != nil && vh.RateLimitPolicy.Global != nil {
+		evh.RateLimits = GlobalRateLimits(vh.RateLimitPolicy.Global.Descriptors)
+	}
+
+	return evh
+}
+
+// Route converts a DAG route to an Envoy route.
+func Route(dagRoute *dag.Route, vhostName string, secure bool, authService *dag.ExtensionCluster) *envoy_route_v3.Route {
```

Comment:
nit: could this be unexported possibly? maybe name to `buildRoute` or something similar, just to disambiguate between this and the other `Route*` helpers (particularly `RouteRoute`)

Reply:
:+1: yeah a few of the helpers can be unexported now, and yeah naming is tough, but `buildRoute` seems sensible. Will make some updates.